contract/services.py

```diff
@@ -1,41 +1,157 @@
 import core
+import json
 
+from django.db.models.query import QuerySet
+from django.contrib.auth.models import AnonymousUser
+from django.core.serializers.json import DjangoJSONEncoder
+from django.forms.models import model_to_dict
 
-@core.comparable
-class PolicyHolderContract(object):
+from contract.models import Contract as ContractModel, \
+    ContractDetails as ContractDetailsModel, \
+    ContractContributionPlanDetails as ContractContributionPlanDetailsModel
 
-    def __init__(self, policy_holder):
-        self.policy_holder = policy_holder
-        pass
+from policyholder.models import PolicyHolderInsuree
+from contribution_plan.models import ContributionPlanBundleDetails
 
-    def __eq__(self, other):
-        return isinstance(other, self.__class__) and self.__dict__ == other.__dict__
+
+def check_authentication(function):
+    def wrapper(self, *args, **kwargs):
+        if type(self.user) is AnonymousUser or not self.user.id:
+            return {
+                "success": False,
+                "message": "Authentication required",
+                "detail": "PermissionDenied",
+            }
+        else:
+            result = function(self, *args, **kwargs)
+            return result
+    return wrapper
+
+
+class Contract(object):
+
+    def __init__(self, user):
+        self.user = user
+
+    @check_authentication
+    def create(self, contract):
+        try:
+            c = ContractModel(**contract)
+            c.state = ContractModel.STATE_DRAFT
+            c.json_ext = json.dumps(_save_json_external(
+                user_id=self.user.id,
+                datetime=core.datetime.datetime.now(),
+                message=""
+            ), cls=DjangoJSONEncoder)
+            c.save(username=self.user.username)
+            uuid_string = str(c.id)
+            # check if the PH is set
+            if "policy_holder_id" in contract:
+                # run services updateFromPHInsuree and Contract Valuation
+                total_amount = self.__services_update_from_ph_contract_valuation(
+                    policy_holder=contract["policy_holder_id"],
+                    contract_id=uuid_string
+                )
+                c.amount_notified = total_amount
+                c.save(username=self.user.username)
+            dict_representation = model_to_dict(c)
+            dict_representation['id'], dict_representation['uuid'] = (str(uuid_string), str(uuid_string))
+        except Exception as exc:
+            return _output_exception(model_name="Contract", method="create", exception=exc)
+        return _output_result_success(dict_representation=dict_representation)
+
+    def __services_update_from_ph_contract_valuation(self, policy_holder, contract_id):
+        cd = ContractDetails(user=self.user)
+        result_ph_insuree = cd.update_from_ph_insuree(contract_details={
```

Comment:
this "update_from_ph_insuree" should be only in create not in this function because with this implementation the user input will be deleted every time you update the value

Reply:
Ok, I will move this into "create".